### PR TITLE
feat(types): Add gRPC Richer Error Model support (RetryInfo)

### PR DESCRIPTION
**Pull request by flemosr**
## Motivation

The [gRPC Richer Error Model][error-handling] is quite useful to send additional feedback to clients, and is supported by many gRPC libraries in other languages.

## Solution

This PR continues the work initiated at [#1068]. It improves the organization of `tonic-types` modules, and adds support for the `RetryInfo` standard error message type to that crate.

[error-handling]: https://www.grpc.io/docs/guides/error
[#1068]: https://github.com/hyperium/tonic/pull/1068

### CODE REVIEW

**Comment by LucioFranco** [comment]
Oh somehow this totally fell off my radar let me review this on monday! sorry about the delay.

File: tonic-types/src/richer_error/mod.rs
Comment (LucioFranco) [inline tonic-types/src/richer_error/mod.rs:30]: Why encode to vec instead of using BytesMut + BufMut and regular `encode`? So that we avoid the vec round trip?

File: tonic-types/src/richer_error/mod.rs
Comment (LucioFranco) [inline tonic-types/src/richer_error/mod.rs:35]: We probably want to seal this trait similar to https://docs.rs/tower-http/latest/src/tower_http/builder.rs.html#45 that way outside users can't implement it and we can add more required fn as we please without making a breaking change.

File: tonic-types/src/richer_error/std_messages/retry_info.rs
Comment (LucioFranco) [inline tonic-types/src/richer_error/std_messages/retry_info.rs:7]: Nit: I like combining these since they have a similar path.

**Review by LucioFranco — CHANGES_REQUESTED** [review]
LGTM overall have a few comments once those are addressed we can merge. Sorry for the delay again!

File: tonic-types/src/richer_error/mod.rs
Comment (flemosr) [reply tonic-types/src/richer_error/mod.rs]: Using `BytesMut` indeed makes much more sense, thanks! It seems to me that using `BufMut` is not necessary in this case but I may be missing something.

File: tonic-types/src/richer_error/mod.rs
Comment (flemosr) [reply tonic-types/src/richer_error/mod.rs]: You are right. I did not know this pattern, quite useful. Thanks!

File: tonic-types/src/richer_error/std_messages/retry_info.rs
Comment (flemosr) [reply tonic-types/src/richer_error/std_messages/retry_info.rs]: I missed that, thanks!

**Comment by flemosr** [comment]
No problem! Hopefully I addressed all comments but please let me know if you think we should make any further changes.

**Review by LucioFranco — APPROVED** [review]
LGTM!

**Pull request merged**
